<?xml version="1.0"?> 
<robot name="upper_limb" xmlns:xacro="http://www.ros.org/wiki/xacro"> 

  <!-- Import all Gazebo-customization elements, including Gazebo colors --> 
  <xacro:include filename="$(find wearable_robot_description)/urdf/upper_limb.gazebo"/> 
  <!-- Import Rviz colors --> 
  <xacro:include filename="$(find wearable_robot_description)/urdf/materials.xacro"/> 

    <link name="world"/>

    <joint name="fixed" type="fixed">
      <parent link="world"/>
      <child link="arm_link"/>
    </joint>

    <link name="arm_link"> 
      <visual> 
        <geometry> 
          <mesh filename="$(find wearable_robot_description)/meshes/upper_limb/arm_link.dae" scale="0.001 0.001 0.001"/> 
        </geometry> 
        <!-- <origin rpy="0 0 -1.570796" xyz="0.0 0 0.0"/>  -->
        <origin rpy="-3.141592 -1.570706 0" xyz="0.1 -0.05 0.00"/>
        <material name="blue"> 
          <color rgba="0 0 1 1"/> 
        </material> 
      </visual> 
      <inertial>
        <mass value="0.1"/>
        <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      </inertial>
    </link>

    <joint name="arm_joint" type="revolute">
        <parent link="arm_link"/>
        <child link="forearm_link"/>
        <origin xyz="0 0 0.1575"/>
        <axis xyz="0 0 0" />
        <limit effort="300" velocity="0.1" lower="-4.1" upper="-1.5"/>
        <dynamics damping="50" friction="1"/>
    </joint>

    <link name="forearm_link">
        <visual>
            <geometry>
                <mesh filename="$(find wearable_robot_description)/meshes/upper_limb/forearm_link.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <!-- <origin rpy="0 -1.570796 -1.570796" xyz="0.0 0.0 0.0"/> -->
            <origin rpy="-1.570796 0 -1.570796" xyz="0.10 0.1575 -0.05"/>
            <material name="red">
                <color rgba="1 0 0 1"/>
            </material>
        </visual>
        <inertial>
            <mass value="0.1"/>
            <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
        </inertial>
    </link>

    <joint name="forearm_band_link_fix_to_forearm_link" type="fixed">
      <!-- <origin xyz="0.018 0.20 0" rpy="0 1.570796 0"/> [> x <] -->
      <origin xyz="0.0 0.22 0" rpy="0 1.570796 0"/> <!-- x -->
      <parent link="forearm_link"/>
      <child link="forearm_band_link"/>
    </joint>

    <link name="forearm_band_link">
      <visual>
        <geometry>
          <!-- <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.03 0.03 0.01"/> -->
          <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.052 0.052 0.01"/>
        </geometry>
        <!-- <origin rpy="1.570796 1.570796 0" xyz="0.05 -0.025 0.0"/> -->
        <origin rpy="1.570796 1.570796 0" xyz="-0.0250 -0.025 0.0"/>
      </visual>
      <collision>
        <geometry>
          <!-- <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.03 0.03 0.01"/> -->
          <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.052 0.052 0.01"/>
        </geometry>
        <!-- <origin rpy="1.570796 1.570796 0" xyz="0.05 -0.025 0.0"/> -->
        <origin rpy="1.570796 1.570796 0" xyz="-0.0250 -0.025 0.0"/>
      </collision>
 
      <inertial>
        <mass value="0.001"/>
        <inertia ixx="1e-2" iyy="1e-2" izz="1e-2" ixy="0" ixz="0" iyz="0"/>
      </inertial>
    </link>

    <joint name="arm_band_link_to_arm_link" type="fixed">
      <origin xyz="0.05 -0.05 0.2875" rpy="0 0 1.570796"/>
      <axis xyz="0 0 0" />
      <parent link="arm_link"/>
      <child link="arm_band_link"/>
    </joint>
       
    <link name="arm_band_link">     
      <visual>
        <geometry>
          <!-- <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.0295 0.0295 0.01"/> -->
          <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.05 0.05 0.01"/>
        </geometry>
        <origin rpy="0 0 0" xyz="0.05 0.01 0.02"/>
      </visual>

      <!-- <collision> -->
      <!--   <geometry> -->
      <!--     [> <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.0295 0.0295 0.01"/> <] -->
      <!--     <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.07 0.07 0.01"/> -->
      <!--   </geometry> -->
      <!--   <origin rpy="0 0 0" xyz="0.05 0.01 0.02"/> -->
      <!-- </collision> -->

      <visual>
        <geometry>
          <!-- <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.0295 0.0295 0.01"/> -->
          <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.05 0.05 0.01"/>
        </geometry>
        <origin rpy="0 0 0" xyz="0.05 0.01 -0.07"/>
      </visual>

      <!-- <collision> -->
      <!--   <geometry> -->
      <!--     [> <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.0295 0.0295 0.01"/> <] -->
      <!--     <mesh filename="$(find wearable_robot_description)/meshes/pipe.dae" scale="0.07 0.07 0.01"/> -->
      <!--   </geometry> -->
      <!--   <origin rpy="0 0 0" xyz="0.05 0.01 -0.07"/> -->
      <!-- </collision> -->

      <inertial>
        <mass value="0.001"/>
        <inertia ixx="1e-2" iyy="1e-2" izz="1e-2" ixy="0" ixz="0" iyz="0"/>
      </inertial>
    </link>

    <!-- band point -->
    <joint name="arm_band_point_1_to_arm_link" type="fixed">
      <origin xyz="0.05 -0.05 0.2875" rpy="0 0 1.570796"/>
      <axis xyz="0 0 0" />
      <parent link="arm_link"/>
      <child link="arm_band_point_1"/>
    </joint>

    <joint name="arm_band_point_2_to_arm_link" type="fixed">
      <origin xyz="0.05 -0.05 0.2875" rpy="0 0 1.570796"/>
      <axis xyz="0 0 0" />
      <parent link="arm_link"/>
      <child link="arm_band_point_2"/>
    </joint>

    <joint name="arm_band_point_3_to_arm_link" type="fixed">
      <origin xyz="0.05 -0.05 0.2875" rpy="0 0 1.570796"/>
      <axis xyz="0 0 0" />
      <parent link="arm_link"/>
      <child link="arm_band_point_3"/>
    </joint>

    <joint name="arm_band_point_4_to_arm_link" type="fixed">
      <origin xyz="0.05 -0.05 0.2875" rpy="0 0 1.570796"/>
      <axis xyz="0 0 0" />
      <parent link="arm_link"/>
      <child link="arm_band_point_4"/>
    </joint>
    
    <link name="arm_band_point_1">    
      <visual>
        <geometry>
          <box size="0.004 0.004 0.004" />
        </geometry>
        <origin rpy="0 0 0" xyz="0.020 0.01 0.02"/>
      </visual>
      <!-- <collision> -->
      <!--   <geometry> -->
      <!--     <box size="0.004 0.004 0.004" /> -->
      <!--   </geometry> -->
      <!--   <origin rpy="0 0 0" xyz="0.020 0.01 0.02"/> -->
      <!-- </collision> -->
    </link>
    
    <link name="arm_band_point_2">
      <visual>
        <geometry>
          <box size="0.004 0.004 0.004" />
        </geometry>
        <origin rpy="0 0 0" xyz="0.080 0.01 0.02"/>
      </visual>
      <!-- <collision> -->
      <!--   <geometry> -->
      <!--     <box size="0.004 0.004 0.004" /> -->
      <!--   </geometry> -->
      <!--   <origin rpy="0 0 0" xyz="0.080 0.01 0.02"/> -->
      <!-- </collision> -->
    </link>
    
    <link name="arm_band_point_3">
      <visual>
        <geometry>
          <box size="0.004 0.004 0.004" />
        </geometry>
        <origin rpy="0 0 0" xyz="0.05 -0.020 0.02"/>
      </visual>
      <!-- <collision> -->
      <!--   <geometry> -->
      <!--     <box size="0.004 0.004 0.004" /> -->
      <!--   </geometry> -->
      <!--   <origin rpy="0 0 0" xyz="0.05 -0.020 0.02"/> -->
      <!-- </collision> -->
    </link>
    
    <link name="arm_band_point_4">     
      <visual>
        <geometry>
          <box size="0.004 0.004 0.004" />
        </geometry>
        <origin rpy="0 0 0" xyz="0.05 0.040 0.02"/>
      </visual>
      <!-- <collision> -->
      <!--   <geometry> -->
      <!--     <box size="0.004 0.004 0.004" /> -->
      <!--   </geometry> -->
      <!--   <origin rpy="0 0 0" xyz="0.05 0.040 0.02"/> -->
      <!-- </collision> -->
    </link>
    
  <ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="arm_joint">
      <command_interface name="position">
        <param name="min">1.79</param>
        <param name="max">6.28</param>
      </command_interface>
      <command_interface name="effort">
        <param name="min">-100.0</param>
        <param name="max">+100.0</param>
      </command_interface>
      <state_interface name="position">
	      <param name="initial_value">1.57079</param>          
      </state_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
  </ros2_control>

</robot>
